dig.go

```diff
@@ -66,13 +68,17 @@ type Container struct {
 
 	// Values that have already been generated in the container.
 	values map[key]reflect.Value
+
+	// Values groups that have already been generated in the container.
+	groups map[key][]reflect.Value
```

Comment:
I recommend map[key]map[reflect.Value]struct{} here.

Reply:
I dislike the reliance that would put on the undefined hash behavior of
`reflect.Value`.

See also, https://play.golang.org/p/rXBlkDiBCq. Two `reflect.Value`s
containing the same value don't actually hash to the same thing.
(Although in this case, that's probably what we want since two separate
constructors should be able to produce the same value if they were
provided separately.)

But generally, it seems safer and more obvious to me to shuffle by hand.